File: Data Structures & Algorithms/count-number-of-islands/submission-1.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:

        coords = [(-1,0), (1,0), (0,-1), (0,1)]

        island = 0

        ROWS = len(grid)
        COLS = len(grid[0])

        def in_bounds(r, c, grid):
            return (
                (0 <= r < ROWS) 
                        and (0 <= c < COLS)
                        )

        def dfs(r, c):
            if not in_bounds(r, c, grid) or grid[r][c] == '0':
                return

            grid[r][c] = "0" # flip the value to prevent double counts
            for dr, dc in coords:
                dfs(r + dr, c + dc)

            
        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == "1":
                    dfs(r, c)
                    island += 1

        return island


        



        

File: Data Structures & Algorithms/flood-fill/submission-1.go
func floodFill(image [][]int, sr int, sc int, color int) [][]int {
    ROWS, COLS := len(image), len(image[0])
    startColor := image[sr][sc]
    if startColor == color{
        return image
    }

    var dfs func(r, c int)
    dfs = func(r, c int){
        if r < 0 || c < 0 || r == ROWS || c == COLS || image[r][c] != startColor{
            return
        }
        image[r][c] = color
        delta := [][]int{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}
        for _, coord := range delta{
            dr := coord[0]
            dc := coord[1]
            dfs(r + dr, c + dc)
        }
    }

    dfs(sr, sc)
    return image
}
